libsolidity/analysis/DeclarationContainer.h

```diff
@@ -57,7 +56,21 @@ class DeclarationContainer
 	bool registerDeclaration(Declaration const& _declaration, ASTString const* _name, langutil::SourceLocation const* _location, bool _invisible, bool _update);
 	bool registerDeclaration(Declaration const& _declaration, bool _invisible, bool _update);
 
-	std::vector<Declaration const*> resolveName(ASTString const& _name, bool _recursive = false, bool _alsoInvisible = false) const;
+	/// Finds all declarations that in the current scope can be referred to using specified name.
+	/// @param _name the name to look for.
+	/// @param _recursive if true and there are no matching declarations in the current container,
+	///        recursively searches the enclosing containers as well.
+	/// @param _alsoInvisible if true, include invisible declaration in the results.
+	/// @param _onlyVisibleAsUnqualifiedNames if true, only include declarations which can be
+	///        referenced outside of the current scope without being qualified with the name of the
+	///        scope. I.e. when they shadow names in the enclosing scope(s). This option is **not**
+	///        passed to the enclosing scope when used together with @a _recursive.
```

Comment:
Do you think we should add an assert for `_recursive` and `_onlyVisibleAsUnqualifiedNames` both being true?

Reply:
I actually did that at first but the assertion fails :) We do call it with both being true and this is the expected behavior unfortunately.

Reply:
Trying to answer your question pulled me down a rabbit hole today and I found a new problem with the PR. So it needs more work unfortunately :(

> Is the comment here correct, then?

Yes, it's correct in that respect because `resolveName()` does set `_onlyVisibleAsUnqualifiedNames` in the recursive call to `false`:
```c++
m_enclosingContainer->resolveName(_name, true, _alsoInvisible, false);
```
The only thing that the assert would change would be not having to document this case. But it turns out that this case does actually happen in practice.

But your comment made me think whether this is really the right behavior. And I think it's not so I changed it to propagate the value of the parameter. It's not a problem in practice though because I think it's not possible to trigger this case it with any currently allowed syntax. 

I tried to find a test case that would behave differently depending on whether we propagate `_onlyVisibleAsUnqualifiedNames` or not and I could not think of any. At first I thought this might be affected:
```solidity
contract C {
    enum E {a, b, c}
    struct S {function (E X) external f; uint E;}
}
```
Turns out it isn't. That's because apparently, unlike a function parameter list, the parameter list of a function type is not a scope. At least the `E` in `E X` is never checked for collision with `uint E`. Hypothetically though, if it was a scope, I think that `resolveName()` should not consider `uint E` from the outer scope a candidate - so `_onlyVisibleAsUnqualifiedNames` should be `true` when recursively visiting the outer scopes as well.

BTW, the comment wasn't really correct in another respect so I have rewritten it a bit.